Refactor VAT handling (Work in Progress)

mamhoff opened the pull request:
This PR should address Issue #6154. 
## This PR breaks existing stores

This PR is _breaking_. It assumes all prices to be entered without tax, which will make _prices wrong_ in stores that have entered their prices including tax. If this has a chance of being accepted, we'd write a rake task that converts prices from gross to net prices for existing stores. 
## New meaning for `Zone.default_tax`

The `default_tax` zone has a slightly different role: It is only there for calculating included tax on orders that don't have a zone assigned yet. This also removes the need for loading the default zone when looking for applicable tax_rates. 
## New meaning for `included_in_tax`

The `included_in_price` flag can now be set even when there is no default zone. It is not used anymore to calculate net prices, but only for tagging VATs as VATs. At a later stage, this flag will be used for displaying prices including VAT not-logged in users.
## New convention: All prices in the database are net prices

Adopting the convention that all prices are entered and calculated upon excluding VAT simplifies VAT handling enormously. This is the reason a lot of code (especially in tax_rate.rb and the default tax calculator) could go away. I took care to adapt Spree's tests to the new situation. 

Like I said, this is still work in progress. Variants are still not shown with the default price added, and there is stuff to do for shipping rates and promotions, too.

To have an idea what this is supposed to accomplish, I set up a separate spec in `core/spec/models/moss_taxes_spec.rb`. These specs fail on every other Spree version. 

Please please please, if anyone has a chance to look over this code or even try it out, I'd be more than grateful for Feedback. Here's a reference to some of the people who have provided notable feedback or code on taxes in issues I've read: @JDutil @peterberkenbosch @vkvelho @jdurand @gmacdougall @GeekOnCoffee  @cduv @radar @Hates @simmsy. Feel free to pull more people into the conversation. 

Right now, you can follow my commit history to see what I did when. For the PR to be merged, I will rebase; and there will be force pushes to this PR at some point. You've been warned :skull:.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://magiclabs-refactor-tax.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://magiclabs-refactor-tax.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

JDutil commented on common_spree_dependencies.rb:
```diff
@@ -3,6 +3,7 @@
 # the one component of Spree.
 source 'https://rubygems.org'
 
+gem 'rails', '4.2.0'
```
spree_core.gemspec does this

tvdeyen replied:
also in `3-0-stable`?

tvdeyen replied:
https://github.com/magiclabs/spree/blob/refactor-tax/core/spree_core.gemspec#L39 also allows 4.2.1 what currently breaks `rspec-activemodel-mocks`

JDutil replied:
Yup 3.0.x was our rails 4.2.x upgrade.  I stole your patch so that shouldn't be an issue:
https://github.com/spree/spree/blob/3-0-stable/core/spec/support/rspec-activemodel-mocks-patch.rb

tvdeyen replied:
There is no stealing in open source ;)

JDutil commented:
I like that this approach would simplify the code base. I feel these changes would be more appropriate as a Spree 4 than 3.1 though as they are breaking.  We've considered doing a 3.1 release with more minor enhancements such as v2 api w/AMS & store credits, but perhaps we should just bump to 4.0 instead once this is ready & skip 3.1.

tvdeyen commented:
:tada: nice work @mamhoff :100:

tvdeyen commented:
@JDutil yes 4.0 would be better I think. Although **we will provide a rake task** for shops currently entering prices including tax. So, this shouldn't be breaking in reality.

JDutil commented:
Please make sure to update release notes with the relevant changes, and instructions on the rake task usage:
https://github.com/spree/spree/blob/master/guides/content/release_notes/3_1_0.md

mamhoff committed:
Spree::Zone.potential_matching_zones does not return default zone

The inclusion of the default zone is only necessary for deducting included
tax from prices that are stored as gross prices. As forthcoming prices will
always be stored net, that won't be necessary. Also removes some complexity.

mamhoff committed:
Increase DB Precision and Scale on all price fields

Previously, Spree used a scale of 2 (2 digits after the comma) on Decimal fields in
the DB. This allows Spree to save values with higher precision.

This is needed for storing prices that are displayed with VAT - only when saved
with high precision, VATs can be calculated correctly even with high (>2)  numbers of items.

mamhoff committed:
Display product prices including VAT

Extend the display_price helper so that it uses the tax zone of the current_oder
to choose a VAT and add that when displaying prices for products.

mamhoff committed:
Delegate form_field_price to default_price

Prices for products with VAT will be shown including VAT on the backend, but
saved excluding VAT in the database. We do this so we can have precise prices
excluding VAT in the DB without making the user calculate those on her own.

mamhoff committed:
Refactor and simplify Spree::TaxRate

In Model:

- Make `Spree::TaxRate.potential_rates_for_zone` a scope

(Ordering is not necessary anymore, since that was only used to deduct default
taxes if that was applicable).

- Remove validator

It is now perfectly possible to enter prices without tax and have VAT calculated on the
fly. We don't need the validator for `Spree::Zone.default_zone` being set.

- Remove `Spree::TaxRate#potentially_applicable`

Without a default zone, all conditions in here are always met by
`Spree::TaxRate.potential_rates_for_zone`.

- Remove complicated logic in `#adjust` and `#compute_amount`

No need since no `default_zone`

- Remove `.store_pre_tax_amount`

If all prices are stored pre tax, we don't need to store them.

In Specs:
- In order to find out an order's tax zone, use `Spree::Order#tax_zone` only.
  Removed stubs for `Spree::Order#tax_address`.

- Adjust expectations so that the default zone doesn't show up for every order.

- Spec `Spree::TaxRate#included_tax_for(zone, category)`

- Remove references to `Spree::TaxRate.store_pre_tax_amount`

- Since there are no default rates anymore, there are no refunds either.

- Since `line_item.price` is now excluding tax, use `line_item.total` instead.

mamhoff committed:
Add display methods that add VAT on the fly to `.money_methods`

These methods (`#display_*_adding_vat` work similarly to those already defined
by Spree::DisplayMoney,  except they add a VAT defined by "included_tax_amount"
(which has to be availableon the calling instance). Otherwise, the amount
displayed will be equal to the one given to `#display_*`.

mamhoff committed:
Include `included_tax_total` in `adjustment_total`

Since prices are stored excluding VAT, now the `included_tax_total`
has to influence the `adjustment_total`.

mamhoff committed:
Override `Spree::Adjustment#amount_adding_vat`

An eventual VAT can only be added if the adjustable is not an order (we cannot
know the tax category for an entire order) or if the Adjustment is a tax adjustment
(That would mean applying the tax twice).

mamhoff committed:
Simplify the default calculator

Since there is no more tax to deduce (all Taxes, including VATs, are added),
the calculator doesn't have to know how to deduce taxes anymore. All code related
to that task is now gone, and the Specs are accordingly adjusted.

This commit also adds a spec for a line item with a very exact price and a
promo adjustment, so we know taxes are correctly calculated for that particular case.

Line items now delegate `#tax_zone` to their order, so that an adjustment can ask for the
correct tax zone for that line item.

mamhoff committed:
Correct indentation on calculator default_refund_amount.rb

mamhoff committed:
Use money_methods to display amounts on line item

For the traditional methods `.*money`, aliases are introduced. Also adds some tests for those.

mamhoff committed:
Override `item_total_adding_vat`

On the order, the item total adding VAT should show the VAT total excluding
Promo adjustments. This override (along with a method on `line_item.rb`)
achieves that.

mamhoff committed:
Add a `#display_price_adding_vat` for `Spree::Price`

It uses the default zone or the order's zone to choose and apply
a VAT to the price of a a variant.

mamhoff committed:
Simplify `Spree::ReimbursementTaxCalculator`

With no taxes to deduce, this one becomes simpler, too.

The specs have been amended, because tax rate now only applies when its zone
matches the orders zone.

mamhoff committed:
Amend the reimbursement spec

Since there is no tax deductions anymore, all reimbursement amounts are positive.
I also removed a rounding in the spec that really should not be in an expectation.

mamhoff committed:
Adjust `Spree::Shipment` to new VAT calculation

Shipment now delegates `tax_zone` to its `order` so that adjustment's prices
can be properly shown.

Also: `pre_tax_amount` will always be the same as `discounted_cost`: Instead of writing to the DB,
it is now aliased to `discounted_cost`.

mamhoff committed:
Refactor `Spree::ShippingRate`

ShippingRate could be reduced in complexity a little as well, since there are no more
shipping rates that include VAT. Besides, I corrected a few expectations (there was one
in particular that was wrong).

mamhoff committed:
Refactor `Spree::Stock::Estimator`

Because the default zone does not have to be included in all tax
calculations, there's no need to load it here. This should address
https://github.com/spree/spree/issues/6234.

mamhoff committed:
Add `TaxCategory#has_vat_rate_for_default_zone?`

mamhoff committed:
`Spree::Product`: Save prices excluding VAT

Since we need net prices to be stored with high scale (five digits min),
and we can't ask our users to be so exact, we let them enter gross prices and convert
before saving to a net price.

mamhoff committed:
Use display methods `display_*_adding_vat` to display prices

End-Users in VAT countries need to be shown prices including VAT by law.

This accomplishes that. For countries that do not have a VAT, the prices are shown
as-is.

mamhoff commented:
The current state of things is: I think this works. Please try it out, and please provide cases where it doesn't. Also: I'm grateful for anyone reviewing this and commenting (don't be shy!). 

Still to do: 

- [ ] Add a rake task for migrating existing stores
- [ ] Allow admin users to enter costs including VAT for shipping rates, promotions, etc.  and automagically save them excluding VAT. 

Notable changes: Increased precision on all price fields, simplified tax rate selection, added `display_[anything]_adding_vat` methods.

JDutil commented:
Please make sure to add updating release notes to your todo list with the relevant changes, and instructions on the rake task usage:
https://github.com/spree/spree/blob/master/guides/content/release_notes/3_1_0.md

alepore commented:
`New convention: All prices in the database are net prices`

Uhm, in my experience this will add complexity for all EU stores.
All prices are always including taxes here.
VAT can change, and sometimes stores doesn't want to update prices after a small VAT change. They just have one price, and that's what the use should pay.
I've done tens of EU store and no one ever talked about net price, they don't known it and don't use it.

Isn't the same in Germany?

alepore commented:
@mamhoff saw your awesome explanation on #6154... pretty complex topic, thanks for that.

My main question is: 
My typical use case is to import prices straight from an external source.
We don't really think much about taxes, i can add a TaxRate just for display purposes, or skip that.
Will i still able to do this simple workflow without the need to convert prices?

kennyadsl commented:
I agree this PR simplfy codebase and it seems the most logic way to handle prices and taxes (great work @mamhoff) but as @alepore we never needed this kind of scenario for our clients. I agree with him that it could create issues for the majority of them while adding product prices. 

Just thinking loud: what about letting (via some preference) choose which kind of price they enter (vat included or not) and remove vat (if included) before write it on the database? I know it could make things much more complex but it's the only  idea I have to support both scenarios without breaking the new proposed convention.

tvdeyen commented:
@alepore @kennyadsl we need net prices for precise calculation. I agree that re-/entering prices in net is unpractical for our customers. That's why we automatically convert VAT prices into net prices after save. So shop owners don't need to enter net prices. We plan to add some JavaScript magic for automatically updating labels on price fields if tax category changes. 

Yes, in Germany (and AFAIK in all countries having VAT) prices must not be displayed without tax. They always have to be displayed including tax (That's the only difference between sales tax and VAT btw). Therefore we added new `display_%{price_type}_adding_vat` methods that display prices including the vat amount of your default tax zone. If the zone of the order is known they **automatically reflect the tax from the customer and we meet the MOSS taxation rules** \o/

So, we tried **many different scenarios** in how we could achieve this complex issue (spent two two week sprints!) and finally came up with this solution. 

The advantage is, that existing stores having sales tax (IMHO the majority of spree users) don't need to do anything after upgrading. For existing VAT stores we will provide rake tasks and a complete documentation that will help with upgrading. 

Thanks for your feedback. Please keep on reviewing and discussing this. 

Happy easter weekend!

tvdeyen commented:
@alepore @kennyadsl I guess you never needed to care about net prices, because you don't sell digital goods together with physical goods, right? 

And the laws changed with 01.01.2015 and the glorious (recognized the irony?) MOSS taxation rules. You now have to display the prices for digital goods **including the VAT of the customers country**. For physical goods it's the same as before: **display the prices including the VAT of the stores country**. 

Spree does some heavy lifting in calculating the net price for VAT stores behind the scenes for you. But, it only calculates the net price for the default tax rate. Besides lots of code for calculating the net price on the fly, that does not work for stores having physical and digital goods. Also spree makes a lot of rounding errors while calculating the net price on the fly and does not do this for all prices (I.e. Shipping costs). 

So, this looks like a big change for existing VAT stores (and it actually is), but is less complexity in calculating and therefore less rounding issues and more readable code.

mamhoff commented:
What @tvdeyen said, generally :) I'd like to add a little bit of detail to both of your questions: 

@alepore We actually modify the price setter to automagically convert the price to a net price if the default zone has VATs before saving to the database, see https://github.com/magiclabs/spree/blob/refactor-tax/core/app/models/spree/product.rb#L80 and https://github.com/magiclabs/spree/blob/refactor-tax/core/app/models/spree/product.rb#L223-L237. If you get gross prices, you should be able to create the products with the gross price and still end up with the net price in the DB. 

@kennyadsl I think this is exactly the kind of solution you're suggesting. There's no preference for it though, because we actually *need* net prices to be exactly calculated (to six digits precision). 

@alepore Your workflow will have to change slightly with this PR if you're in Europe. You'll want to create a default zone and TaxRate that has `included_in_price` set before importing prices. 

@alepore If VATs change (which they do really seldomly, as that is politically a little loaded...), and your shop does not adjust its gross prices, they *are* changing their prices. I think it's a good thing for merchants to be aware of that kind of change as well. If you want to keep prices as before - just change them back to the previous price afterwards.

alepore commented:
@tvdeyen @mamhoff thanks for all the effort on this, i'm confident you're getting this right!

anyway, i understand the whole logic should also work if i don't create a default tax rate, right?
probably not the best idea but can make sense for simple/no digital goods scenarios..

mamhoff commented:
@alepore If you don't create a default tax rate, Spree will assume the prices you enter as net prices. If you add a tax rate after entering product prices, your prices will be inflated by your VAT (because at the time of entering prices, Spree has no knowledge of how much VAT to deduct). 

If your VAT rules are anything similar to the German ones though, I *strongly* recommend setting VAT rates correctly at the start, because only then will you be able to display all the legally required "VAT included" messages...

alepore commented:
Ok, that's clearly an hack.
Maybe a warning or a validation should avoid this error?

mamhoff commented:
We are planning on changing the label on the price input field to say "net price" or "gross price" depending on whether the product's tax category has a default VAT. 

We can't add a warning when you enter a price when there isn't a default VAT, because that's the standard use case for Sales tax countries. 

I can imagine a warning when you change the default zone's VAT rate, and re-introducing the "you have to choose a default zone" validation on adding the `included_in_price` boolean. However, I think documenting it is better than warning though: https://github.com/spree/spree-guides/blob/master/content/developer/core/taxation.md will have to be adapted and amended with a "Setting up Spree for VAT taxation" guide. 

As for the "Hack" claim - if you have a better suggestion, I'd be absolutely thrilled to hear about it :)

alepore commented:
Oh yeah, that's actually not a hack for sales tax countries, got it :+1: 
Sorry for the confusion :)

mamhoff commented:
Well, well. So it turned out that net prices are *not* the right way to go - if gross prices change (as is the case with MOSS (#6154, the issue that started it all), their net prices actually *do* change with them. 

I will close this PR and open a new one that has a solution that
 - does not break existing stores
 - still simplifies some of Spree's tax handling (although not as much, as we still need to deduct prices and store pre-tax-amounts

It is PR #6293 .
